ext/datadog_cov/datadog_cov.c

```diff
@@ -0,0 +1,177 @@
+#include <ruby.h>
+#include <ruby/debug.h>
+
+// Utils
+static int prefix(const char *pre, const char *str)
+{
+  return strncmp(pre, str, strlen(pre));
+}
+
+// const
+#define DD_COV_TARGET_FILES 1
+#define DD_COV_TARGET_LINES 2
+
+// Data structure
+struct dd_cov_data
+{
+  char *root;
+  int mode;
+  VALUE coverage;
+};
+
+static void dd_cov_mark(void *ptr)
+{
+  struct dd_cov_data *dd_cov_data = ptr;
+  rb_gc_mark_movable(dd_cov_data->coverage);
+}
+
+static void dd_cov_free(void *ptr)
+{
+  struct dd_cov_data *dd_cov_data = ptr;
+
+  xfree(dd_cov_data);
+}
+
+static void dd_cov_compact(void *ptr)
+{
+  struct dd_cov_data *dd_cov_data = ptr;
+  dd_cov_data->coverage = rb_gc_location(dd_cov_data->coverage);
+}
+
+const rb_data_type_t dd_cov_data_type = {
+    .wrap_struct_name = "dd_cov",
+    .function = {
+        .dmark = dd_cov_mark,
+        .dfree = dd_cov_free,
+        .dsize = NULL,
+        .dcompact = dd_cov_compact},
+    .flags = RUBY_TYPED_FREE_IMMEDIATELY};
+
+static VALUE dd_cov_allocate(VALUE klass)
+{
+  struct dd_cov_data *dd_cov_data;
+  VALUE obj = TypedData_Make_Struct(klass, struct dd_cov_data, &dd_cov_data_type, dd_cov_data);
+  dd_cov_data->coverage = rb_hash_new();
+  return obj;
+}
+
+// DDCov methods
+static VALUE dd_cov_initialize(int argc, VALUE *argv, VALUE self)
+{
+  VALUE opt;
+  int mode;
+
+  rb_scan_args(argc, argv, "10", &opt);
+  VALUE rb_root = rb_hash_lookup(opt, ID2SYM(rb_intern("root")));
+  if (!RTEST(rb_root))
+  {
+    rb_raise(rb_eArgError, "root is required");
+  }
+
+  VALUE rb_mode = rb_hash_lookup(opt, ID2SYM(rb_intern("mode")));
+  if (!RTEST(rb_mode) || rb_mode == ID2SYM(rb_intern("files")))
+  {
+    mode = DD_COV_TARGET_FILES;
+  }
+  else if (rb_mode == ID2SYM(rb_intern("lines")))
+  {
+    mode = DD_COV_TARGET_LINES;
+  }
+  else
+  {
+    rb_raise(rb_eArgError, "mode is invalid");
+  }
+
+  struct dd_cov_data *dd_cov_data;
+  TypedData_Get_Struct(self, struct dd_cov_data, &dd_cov_data_type, dd_cov_data);
+
+  dd_cov_data->root = StringValueCStr(rb_root);
+  dd_cov_data->mode = mode;
+
+  return Qnil;
+}
+
+static void dd_cov_update_line_coverage(rb_event_flag_t event, VALUE data, VALUE self, ID id, VALUE klass)
+{
+  struct dd_cov_data *dd_cov_data;
+  TypedData_Get_Struct(data, struct dd_cov_data, &dd_cov_data_type, dd_cov_data);
+
+  const char *filename = rb_sourcefile();
+  if (filename == 0)
+  {
+    return;
+  }
+
+  if (prefix(dd_cov_data->root, filename) != 0)
+  {
+    return;
+  }
```

Comment:
Not sure if caching filename will make it faster: then I will need to check if `rb_source_file()` returned the same file name as `cached_filename` and this will require another call to `strcmp` so I win nothing in the best case. Is there something I miss?

Reply:
I was thinking that you may be getting a pointer into the same string so if that's the case you could do a `==` check.

If it changes it may not be a win as you point out, although the cost of the `strcmp` will be really low as the previous line will quite likely be in the cpu cache anyway.